packages/web3torrent/src/clients/web3t-channel-client.ts

```diff
@@ -0,0 +1,193 @@
+      allocations,
+      appDefinition,
+      'appData unused'
+    );
+
+    return convertToChannelState(channelResult);
+  }
+
+  async getAddress() {
+    this.mySigningAddress = await this.channelClient.getAddress();
+    return this.mySigningAddress;
+  }
+
+  async getEthereumSelectedAddress() {
+    this.myEthereumSelectedAddress = await this.channelClient.getEthereumSelectedAddress();
+    return this.myEthereumSelectedAddress;
+  }
+
+  onMessageQueued(callback: (message: Message) => void) {
+    return this.channelClient.onMessageQueued(callback);
+  }
+
+  // Accepts an web3t-friendly callback, performs the necessary encoding, and subscribes to the channelClient with an appropriate, API-compliant callback
+  onChannelUpdated(web3tCallback: (channelState: ChannelState) => any) {
+    function callback(channelResult: ChannelResult): any {
+      web3tCallback(convertToChannelState(channelResult));
+    }
+    // These are two distinct events from the channel client
+    // but for our purposes we can treat them the same
+    // and rely on the channel status
+    const unsubChannelUpdated = this.channelClient.onChannelUpdated(callback);
+    const unsubChannelProposed = this.channelClient.onChannelProposed(callback);
+
+    return () => {
+      unsubChannelUpdated();
+      unsubChannelProposed();
+    };
+  }
+
+  async joinChannel(channelId: string) {
+    const channelResult = await this.channelClient.joinChannel(channelId);
+    return convertToChannelState(channelResult);
+  }
+
+  async closeChannel(channelId: string): Promise<ChannelState> {
+    const channelResult = await this.channelClient.closeChannel(channelId);
+    return convertToChannelState(channelResult);
+  }
+
+  async challengeChannel(channelId: string): Promise<ChannelState> {
+    const channelResult = await this.channelClient.challengeChannel(channelId);
+    return convertToChannelState(channelResult);
+  }
+
+  async updateChannel(
+    channelId: string,
+    aAddress: string,
+    bAddress: string,
+    aBal: string,
+    bBal: string,
+    aOutcomeAddress: string = aAddress,
+    bOutcomeAddress: string = bAddress
+  ) {
+    const allocations = formatAllocations(aOutcomeAddress, bOutcomeAddress, aBal, bBal);
+    const participants = formatParticipants(aAddress, bAddress, aOutcomeAddress, bOutcomeAddress);
+
+    // ignore return val for now and stub out response
+    const channelResult = await this.channelClient.updateChannel(
+      channelId,
+      participants,
+      allocations,
+      'appData unused'
+    );
+
+    return convertToChannelState(channelResult);
+  }
+
+  // seeder may use this method to accept payments
+  async counterSign(channelState: ChannelState) {
+    const {
+      channelId,
+      aAddress,
+      bAddress,
+      aBal,
+      bBal,
+      aOutcomeAddress,
+      bOutcomeAddress
+    } = channelState;
+    this.updateChannel(channelId, aAddress, bAddress, aBal, bBal, aOutcomeAddress, bOutcomeAddress);
+  }
```

Comment:
Something we could do with Web3Torrent is the following:

- Use the null app
- Sender (leecher) sends state with the same `turnNum` to the Receiver (seeder) with `isFinal` set to `true`
- Receiver stores all of these inside their wallet

This gives the Receiver the option to close the channel unilaterally at any point in time by countersigning the state with `isFinal` as `true` on whichever is the most preferred to them. The client in Web3Torrent would select the most optimal one based on `aBal`/`bBal` and call a method on the wallet.

This would require the wallet to have some logic about "selecting" a state if there are multiple of them with the same turn number.

Reply:
I think for now we should target the way the wallet currently works and then switch to something more optimal later on.